src/main/java/ru/ralsei/whatcanyousee/gameactivity/AudioConnector.java

```diff
@@ -0,0 +1,145 @@
+package ru.ralsei.whatcanyousee.gameactivity;
+
+import android.media.AudioAttributes;
+import android.media.AudioFormat;
+import android.media.AudioManager;
+import android.media.AudioRecord;
+import android.media.AudioTrack;
+import android.media.MediaRecorder;
+import android.util.Log;
+
+import com.google.android.gms.games.multiplayer.Participant;
+
+/**
+ * Class for managing voice connection between players.
+ */
+class AudioConnector {
+    private String TAG = "What can you see: audio connector";
+
+    private GameActivity activity;
+
+    AudioConnector(GameActivity activity) {
+        this.activity = activity;
+    }
+
+    /**
+     * Minimum buffer size to record audio.
+     */
+    private final int MIN_BUFFER_SIZE = AudioRecord.getMinBufferSize(8000, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT);
+
+    /**
+     * Object for streaming audio received as byte buffer.
+     */
+    private AudioTrack track;
+
+    /**
+     * True if player broadcasts his voice, if false --- voice recording pauses.
+     */
+    private volatile boolean broadcastAudio = false;
+
+    /**
+     * Thread for broadcasting audio.
+     */
+    private Thread broadcastThread;
+
+    /**
+     * Voice recorder that translate input voice into byte buffer.
+     */
+    private AudioRecord recorder;
+
+    /**
+     * Prepares AudioTrack.
+     */
+    void prepareReceiveAudio() {
+        track = new AudioTrack(
+                new AudioAttributes.Builder()
+                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
+                        .build(),
+                new AudioFormat.Builder()
+                        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
+                        .setSampleRate(8000)
+                        .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
+                        .build(), MIN_BUFFER_SIZE *10, AudioTrack.MODE_STREAM, AudioManager.AUDIO_SESSION_ID_GENERATE);
+        track.play();
+
+        Log.d(TAG, "Audio reception prepared");
+    }
+
+    /**
+     * Initialise audio recorder and starts thread that sends recorded audio to other player infinitely.
+     */
+    void prepareBroadcastAudio() {
+        if (activity.getGooglePlayHandler().getTeammateParticipant() == null || activity.getGooglePlayHandler().getTeammateParticipant().getStatus() != Participant.STATUS_JOINED) {
+            activity.getUIHandler().showGameError();
+            activity.getGooglePlayHandler().leaveRoom();
+        }
+
+        recorder = new AudioRecord(MediaRecorder.AudioSource.MIC, 8000, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT, MIN_BUFFER_SIZE * 10);
+
+        broadcastThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                byte[] buffer = new byte[GooglePlayHandler.MAX_RELIABLE_BUFFER_SIZE];
+
+                recorder.startRecording();
+                while (!Thread.interrupted()) {
```

Comment:
Done, but: "clearResources would have to use a similar mechanism" --- broadcastThread.interrupt() would do the trick, wouldn't it?

Reply:
Oh, that one gave me some hard times to understand: I totally forgot about the inner while, that break on l.88 supposed to exit the entire thread (break -> return). Writing this in any other way seems to me a complication though. A bit hard to test, btw, because I never actually made a button for stopping recording audio, but it supposed to be...

There was another problem with broadcastAudio flag not clearing on clearResources, so starting the next game has been failing, but that one I just caught.

Reply:
Oh, also, I was stopping AudioRecord on a wrong thread. Now it all seems like working.

Reply:
Okay, seems good. I was referring to writing a unit test though (which you can run 1000 times to make sure the threads do stop no matter the order of execution).